File: 2104. Sum of Subarray Ranges.py
# 2104. Sum of Subarray Ranges 

class Solution:
    def subArrayRanges(self, nums: List[int]) -> int:
        n, answer = len(nums), 0
        stack = []

        #
        for right in range(n+1):
            while stack and (right == n or nums[stack[-1]] >= nums[right]):
                mid = stack.pop()
                left = -1 if not stack else stack[-1]
                answer -= nums[mid] * (mid - left) * (right - mid)
            stack.append(right)
        
        #
        stack.clear()
        for right in range(n+1):
            while stack and (right == n or nums[stack[-1]] <= nums[right]):
                mid = stack.pop()
                left = -1 if not stack else stack[-1]
                answer += nums[mid] * (mid-left) * (right-mid)
            stack.append(right)

        return answer


        # small_sum = 0
        # large_sum = 0
        
        # stack = [-1]
        # nums.append(-float('inf'))
        # for i, n in enumerate(nums):
        #     while nums[stack[-1]] > n:
        #         popInd = stack.pop()
        #         val = nums[popInd]
        #         small_sum += (popInd - stack[-1]) * (i - popInd) * val
        #     stack.append(i)

        # stack = [-1]
        # nums[-1] = float('inf')
        # for i, n in enumerate(nums):
        #     while nums[stack[-1]] < n:
        #         popInd = stack.pop()
        #         val = nums[popInd]
        #         large_sum += (popInd - stack[-1]) * (i - popInd) * val
        #     stack.append(i)
        
        # return large_sum - small_sum

File: 2104. Sum of Subarray Ranges.cpp
//2104. Sum of Subarray Ranges

#include <vector>
#include <stack>

using namespace std;

class Solution {
    public:
        long long subArrayRanges(vector<int>& nums) {
            long long ans {0};
            int n = nums.size();
            stack<int> sm_st, lg_st;
    
            for (int right = 0; right < n+1; right++) {
                while (!sm_st.empty() && (right == n || nums[sm_st.top()] > nums[right])) {
                    int mid = sm_st.top();
                    sm_st.pop();
                    int left = !sm_st.empty() ? sm_st.top() : -1;
                    ans -= (long long)(right - mid) * (mid - left) * nums[mid];
                }
                sm_st.push(right);
            }
    
            for (int right = 0; right < n+1; right++) {
                while (!lg_st.empty() && (right == n || nums[lg_st.top()] < nums[right])) {
                    int mid = lg_st.top();
                    lg_st.pop();
                    int left = !lg_st.empty() ? lg_st.top() : -1;
                    ans += (long long)(right - mid) * (mid - left) * nums[mid];
                }
                lg_st.push(right);
            }
            return ans;
        }
    };
    
    